File: _includes/code/howto/configure.schema.ts
// How-to: Configure -> Schema - TypeScript examples
// run with: node --loader=ts-node/esm configure.schema.ts
import assert from 'assert';

// ================================
// ===== INSTANTIATION-COMMON =====
// ================================
import weaviate from 'weaviate-ts-client';

const client = weaviate.client({
  scheme: 'http',
  host: 'localhost:8080',
  headers: {
    'X-OpenAI-Api-Key': process.env['OPENAI_API_KEY'],
  },
});

const className = 'Article';

// Clean slate
try {
  await client.schema.classDeleter().withClassName(className).do();
} catch (e) {
  // ignore error if class doesn't exist
}

// START CreateCollection
const emptyClassDefinition = {
  class: 'Article',
};

// Add the class to the schema
let result = await client
  .schema
  .classCreator()
  .withClass(emptyClassDefinition)
  .do();

// The returned value is the full class definition, showing all defaults
console.log(JSON.stringify(result, null, 2));
// END CreateCollection

// Test
console.assert('invertedIndexConfig' in result);

// Delete the class to recreate it
await client.schema.classDeleter().withClassName(className).do();

// START PropertyDefinition
const classWithProps = {
  class: 'Article',
  // highlight-start
  properties: [
    {
      name: 'title',
      dataType: ['text'],
    },
    {
      name: 'body',
      dataType: ['text'],
    },
  ],
  // highlight-end
};

// Add the class to the schema
result = await client
  .schema
  .classCreator()
  .withClass(classWithProps)
  .do();

// The returned value is the full class definition, showing all defaults
console.log(JSON.stringify(result, null, 2));
// END PropertyDefinition

// Test
assert('invertedIndexConfig' in result);

// Delete the class to recreate it
await client.schema.classDeleter().withClassName(className).do();

// START Vectorizer
const classWithVectorizer = {
  class: 'Article',
  properties: [
    {
      name: 'title',
      dataType: ['text'],
    },
  ],
  // highlight-start
  vectorizer: 'text2vec-openai',  // this could be any vectorizer
  // highlight-end
};

// Add the class to the schema
result = await client
  .schema
  .classCreator()
  .withClass(classWithVectorizer)
  .do();

// The returned value is the full class definition, showing all defaults
console.log(JSON.stringify(result, null, 2));
// END Vectorizer

// Test
assert.equal(result['vectorizer'], 'text2vec-openai');
assert.equal(result['properties'].length, 1);  // no 'body' from the previous example

// Delete the class to recreate it
await client.schema.classDeleter().withClassName(className).do();


// START SetVectorIndex
const classWithIndexType = {
  class: 'Article',
  properties: [
    {
      name: 'title',
      dataType: ['text'],
    },
  ],
  vectorizer: 'text2vec-openai',  // this could be any vectorizer
  // highlight-start
  vectorIndexType: 'flat',  // or `hnsw`
  vectorIndexConfig: {
    bq: {
        enabled: true,  // Enable BQ compression. Default: False
        rescoreLimit: 200,  // The minimum number of candidates to fetch before rescoring. Default: -1 (No limit)
        cache: true,  // Enable use of vector cache. Default: False
    },
    vectorCacheMaxObjects: 100000,  // Cache size if `cache` enabled. Default: 1000000000000
  }
  // highlight-end
};

// Add the class to the schema
result = await client
  .schema
  .classCreator()
  .withClass(classWithIndexType)
  .do();

// The returned value is the full class definition, showing all defaults
console.log(JSON.stringify(result, null, 2));
// END SetVectorIndex

// Test
assert.equal(result['vectorizer'], 'text2vec-openai');
assert.equal(result['vectorIndexType'], 'flat');
assert.equal(result['properties'].length, 1);  // no 'body' from the previous example

// Delete the class to recreate it
await client.schema.classDeleter().withClassName(className).do();


// START ModuleSettings
const classWithModuleSettings = {
  class: 'Article',
  properties: [
    {
      name: 'title',
      dataType: ['text'],
    },
  ],
  vectorizer: 'text2vec-cohere',  // this could be any vectorizer
  // highlight-start
  moduleConfig: {
    'text2vec-cohere': {  // this must match the vectorizer used
      vectorizeClassName: true,
      model: 'embed-multilingual-v2.0',
    },
  },
  // highlight-end
};

// Add the class to the schema
result = await client
  .schema
  .classCreator()
  .withClass(classWithModuleSettings)
  .do();

// The returned value is the full class definition, showing all defaults
console.log(JSON.stringify(result, null, 2));
// END ModuleSettings

// Test
assert.equal(result.vectorizer, 'text2vec-cohere');
assert.equal(result.moduleConfig['text2vec-cohere']['model'], 'embed-multilingual-v2.0');

// Delete the class to recreate it
await client.schema.classDeleter().withClassName(className).do();


// START PropModuleSettings
const classWithPropModuleSettings = {
  class: 'Article',
  vectorizer: 'text2vec-huggingface',  // this could be any vectorizer
  properties: [
    {
      name: 'title',
      dataType: ['text'],
      // highlight-start
      moduleConfig: {
        'text2vec-huggingface': {  // this must match the vectorizer used
          vectorizePropertyName: true,
          tokenization: 'lowercase'  // Use "lowercase" tokenization
        },
      },
      // highlight-end
    },
    {
      name: 'body',
      dataType: ['text'],
      // highlight-start
      moduleConfig: {
        'text2vec-huggingface': {  // this must match the vectorizer used
          skip: true,  // Don't vectorize this property
          tokenization: 'whitespace'  // Use "whitespace" tokenization
        },
      },
      // highlight-end
    },
  ],
};

// Add the class to the schema
result = await client
  .schema
  .classCreator()
  .withClass(classWithPropModuleSettings)
  .do();

// The returned value is the full class definition, showing all defaults
console.log(JSON.stringify(result, null, 2));
// END PropModuleSettings

// Test
assert.equal(result.vectorizer, 'text2vec-huggingface');
assert.equal(result.properties[0].moduleConfig['text2vec-huggingface']['vectorizePropertyName'], false);

// Delete the class to recreate it
await client.schema.classDeleter().withClassName(className).do();


// START IndexReplicationSettings
const classWithIndexReplication = {
  class: 'Article',
  // highlight-start
  vectorIndexConfig: {
    distance: 'cosine',
  },
  replicationConfig: {
    factor: 3,
  },
  // highlight-end
};

// Add the class to the schema
result = await client
  .schema
  .classCreator()
  .withClass(classWithIndexReplication)
  .do();

// The returned value is the full class definition, showing all defaults
console.log(JSON.stringify(result, null, 2));
// END IndexReplicationSettings

// Test
assert.equal(result.replicationConfig.factor, 3);

// TODOv4 add the missing example
// START Multi-tenancy
await client.schema
  .classCreator().withClass({
    class: 'Article',
    multiTenancyConfig: { enabled: true },
  })
  .do();
// END Multi-tenancy


// START AddProp
const prop = {
  name: 'body',
  dataType: ['text'],
};

const resultProp = await client
  .schema
  .propertyCreator()
  .withClassName('Article')
  .withProperty(prop)
  .do();

// The returned value is full property definition
console.log(JSON.stringify(resultProp, null, 2));
// END AddProp

// Test
assert.equal(resultProp.name, 'body');


// START ModifyParam
// Not available yet - https://github.com/weaviate/typescript-client/issues/64
// END ModifyParam


// START SchemaGet
const schema = await client
  .schema
  .getter()
  .do();

// The returned value is the entire schema
console.log(JSON.stringify(schema, null, 2));
// END SchemaGet

// Test
assert.equal(schema.classes[0].class, 'Article');

File: _includes/code/howto/configure.schema.py
# How-to: Configure -> Schema - Python examples
import os

# ================================
# ===== INSTANTIATION-COMMON =====
# ================================

import weaviate

# Instantiate the client with the OpenAI API key
client = weaviate.connect_to_local(
    additional_headers={
        "X-OpenAI-Api-Key": os.environ["OPENAI_API_KEY"]
    }
)

# Clean slate
if client.collections.exists("Article"):
    client.collections.delete("Article")


# START CreateCollection
client.collections.create("Article")
# END CreateCollection

# Test
assert (client.collections.exists("Article"))

# Delete the collection to recreate it
client.collections.delete("Article")


# START PropertyDefinition
# highlight-start
import weaviate.classes as wvc
# highlight-end

client.collections.create(
    "Article",
    # highlight-start
    properties=[
        wvc.Property(name="title", data_type=wvc.DataType.TEXT),
        wvc.Property(name="body", data_type=wvc.DataType.TEXT),
    ]
    # highlight-end
)
# END PropertyDefinition

# Test
collection = client.collections.get("Article")
config = collection.config.get()
assert (client.collections.exists("Article"))
assert len(config.properties) == 2


# Delete the collection to recreate it
client.collections.delete("Article")


# START Vectorizer
import weaviate.classes as wvc

client.collections.create(
    "Article",
    # highlight-start
    vectorizer_config=wvc.Configure.Vectorizer.text2vec_openai(),
    # highlight-end
    properties=[ # properties configuration is optional
        wvc.Property(name="title", data_type=wvc.DataType.TEXT),
        wvc.Property(name="body", data_type=wvc.DataType.TEXT),
    ]
)
# END Vectorizer

# Test
collection = client.collections.get("Article")
config = collection.config.get()
assert config.vectorizer.value == "text2vec-openai"

# Delete the collection to recreate it
client.collections.delete("Article")


# START ModuleSettings
import weaviate.classes as wvc

client.collections.create(
    "Article",
    # highlight-start
    vectorizer_config=wvc.Configure.Vectorizer.text2vec_cohere(
        model="embed-multilingual-v2.0",
        vectorize_class_name=True
    ),
    # highlight-end
)
# END ModuleSettings

# Test
collection = client.collections.get("Article")
config = collection.config.get()
assert config.vectorizer.value == "text2vec-cohere"
# TODOv4 make sure we can verify the model name
# assert config.vectorizer.model == "embed-multilingual-v2.0"

# Delete the collection to recreate it
client.collections.delete("Article")


# START PropModuleSettings
import weaviate.classes as wvc

client.collections.create(
    "Article",
    vectorizer_config=wvc.Configure.Vectorizer.text2vec_huggingface(),

    properties=[
        wvc.Property(
            name="title",
            data_type=wvc.DataType.TEXT,
            # highlight-start
            vectorize_property_name=True,  # Use "title" as part of the value to vectorize
            tokenization=wvc.Tokenization.LOWERCASE  # Use "lowecase" tokenization
            # highlight-end
        ),
        wvc.Property(
            name="body",
            data_type=wvc.DataType.TEXT,
            # highlight-start
            skip_vectorization=True,  # Don't vectorize this property
            tokenization=wvc.Tokenization.WHITESPACE  # Use "whitespace" tokenization
            # highlight-end
        ),
    ]
)
# END PropModuleSettings

# Test
collection = client.collections.get("Article")
config = collection.config.get()

assert config.vectorizer.value == "text2vec-huggingface"

# assert result["properties"][0]["moduleConfig"]["text2vec-huggingface"]["vectorizePropertyName"] is False

# Delete the collection to recreate it
client.collections.delete("Article")


# START IndexReplicationSettings
import weaviate.classes as wvc

client.collections.create(
    "Article",
    # highlight-start
    vector_index_config=wvc.Configure.vector_index(
        distance_metric=wvc.VectorDistance.COSINE
    ),
    # highlight-end

    # highlight-start
    replication_config=wvc.Configure.replication(
        factor=3
    )
    # highlight-end
)
# END IndexReplicationSettings

# Test
collection = client.collections.get("Article")
config = collection.config.get()
assert config.vector_index_config.distance_metric.value == "cosine"

# Delete the collection to recreate it
client.collections.delete("Article")


# START Multi-tenancy
client.collections.create(
    "Article",
    # highlight-start
    multi_tenancy_config=wvc.Configure.multi_tenancy(True)
    # highlight-end
)
# END Multi-tenancy


# START AddProp
import weaviate.classes as wvc

# Get the Article collection object
articles = client.collections.get("Article")

# Add a new property
articles.config.add_property(
    # highlight-start
    additional_property=wvc.Property(
        name="body",
        data_type=wvc.DataType.TEXT
    )
    # highlight-end
)
# END AddProp

# Test
collection = client.collections.get("Article")
config = collection.config.get()
assert len(config.properties) == 1
assert config.properties[0].name == "body"


# TODOv4 we need to update this part of the example
# It seemed to be broken when I worked on this example
# stopwords_preset="en",

# START ModifyParam
import weaviate.classes as wvc

# Get the Article collection object
articles = client.collections.get("Article")

# Update the collection configuration
# highlight-start
articles.config.update(
    # Note, use Reconfigure here (not Configure)
    inverted_index_config=wvc.Reconfigure.inverted_index(
        stopwords_removals=["a", "the"]
    )
)
# highlight-end
# END ModifyParam

# Test
collection = client.collections.get("Article")
config = collection.config.get()
assert config.inverted_index_config.stopwords.removals == ["a", "the"]

# DON'T delete the collection yet

# START SchemaGet
collection = client.collections.get("Article")
# highlight-start
config = collection.config.get()
# highlight-end

# print some of the config properties
print(config.vectorizer)
print(config.inverted_index_config)
print(config.inverted_index_config.stopwords.removals)
print(config.multi_tenancy_config)
print(config.vector_index_config)
print(config.vector_index_config.distance_metric)
# END SchemaGet

# Delete the collection to recreate it
client.collections.delete("Article")